src/modules/flight_mode_manager/tasks/ManualAltitude/FlightTaskManualAltitude.cpp

```diff
@@ -229,28 +225,18 @@ void FlightTaskManualAltitude::_respectMaxAltitude()
 {
 	if (PX4_ISFINITE(_dist_to_bottom)) {
 
-		// if there is a valid maximum distance to ground, linearly increase speed limit with distance
-		// below the maximum, preserving control loop vertical position error gain.
-		// TODO: manipulate the velocity setpoint instead of tweaking the saturation of the controller
-		if (PX4_ISFINITE(_max_distance_to_ground)) {
-			_constraints.speed_up = math::constrain(_param_mpc_z_p.get() * (_max_distance_to_ground - _dist_to_bottom),
-								-_param_mpc_z_vel_max_dn.get(), _param_mpc_z_vel_max_up.get());
-
-		} else {
-			_constraints.speed_up = _param_mpc_z_vel_max_up.get();
-		}
+		if ((-_position_setpoint(2) + _position(2) > _max_distance_to_ground - _dist_to_bottom
+		     || -_velocity_setpoint(2) > _max_distance_to_ground - _dist_to_bottom
```

Comment:
this is "1 second away from reaching `_max_distance_to_ground`, right? Is this really needed?

Reply:
What about using this: https://github.com/PX4/PX4-Autopilot/blob/e194a52907083783f7e67ff93504016ca1275ce8/src/lib/mathlib/math/TrajMath.hpp#L48-L71 (as used in collision prevention).